### PR TITLE
Ch5 issue18

### LINKED ISSUE #18: What are the processes involved to build a basic end-to-end workflow, and how do I enhance the workflow with quality-of-life improvements from there?
**Section Outline**
- Processes of a typical end-to-end workflow
- How basic can you get the workflow to be
- How to enhance the workflow with quality-of-life improvements from there
 
**Definition of Done**
- [ ] Completed `5-data-mgmt-exp-proc/e2e-workflow.md`
- [ ] Verified against [self-review checklist](https://docs.google.com/document/d/1jIFbca53uD8b2FsDrUR6Dlb7k20SlZQBt-jpJ_qV14Y/edit)
- [ ] Page builds and displays correctly

### PR DESCRIPTION
**Pull request by Syakyr**

### CODE REVIEW

**Comment by chuawjk** [comment]
Thanks for this @Syakyr. I think this is a really important contribution as it gives a comprehensive outline of the technical components of the an AI project.

1. There's a lot of useful content covered here. In keeping with the short-and-sweet intention of the handbook, I'd like to suggestion splitting this into two sections based your basic and advanced workflow diagrams. This way, the sections can separately target new and experienced engineers, respectively.

2. For version control, I was wondering if you'd like to introduce the concept of data versioning (that being said, I understand that there are word count constraints)?

3. Also under version control, I'd suggest omitting the following until we have a standalone article on version control.
```
Refer to another part of this handbook for more information regarding this process.
```

4. The subsection on developer workspace might also benefit from an explicit definition of what a developer workspace is, and what it is used for. The term is quite broad, so I think it will be useful to align readers on exactly what is meant in this article.

5. Could I trouble you to rephrase the section title as a question? The original idea behind the handbook was a Q&A style. We can use something similar to this issue title with any necessary adjustments.

6. Lastly, I was wondering if this section might be a better fit for another section. The current chapter was intended to cover DataOps and EDA process, but this section provides a broader view. My suggestions is to move it to Chapter 3 Collaborative Development Platforms, but I'm aware that's not a perfect fit either. What do you think?

For your consideration (:

**Comment by Syakyr** [comment]
> I'd like to suggestion splitting this into two sections based your basic and advanced workflow diagrams. This way, the sections can separately target new and experienced engineers, respectively.

Done 👍🏼 

> For version control, I was wondering if you'd like to introduce the concept of data versioning (that being said, I understand that there are word count constraints)?

I've added a small part within the version control section for data versioning, and refer to another website for now, unless we have a standalone article on data versioning.

> I'd suggest omitting "Refer to another part of this handbook for more information regarding this process." until we have a standalone article on version control.

Done 👍🏼 

> 4. The subsection on developer workspace might also benefit from an explicit definition of what a developer workspace is, and what it is used for. The term is quite broad, so I think it will be useful to align readers on exactly what is meant in this article.

Added a small part with regards to this section to point to the existence of remote developer workspaces.

> Could I trouble you to rephrase the section title as a question?

Done 👍🏼 

> I was wondering if this section might be a better fit for another section. The current chapter was intended to cover DataOps and EDA process, but this section provides a broader view. My suggestions is to move it to Chapter 3 Collaborative Development Platforms, but I'm aware that's not a perfect fit either. What do you think?

This could be moved to Chapter 2 on Project Management & Technical Leadership instead, though not sure what user story this would fall under.

**Comment by ngkh2018** [comment]
Basic Workflow

1. A hyperlink seems to be missing in this sentence (‘You can check this section for more information on data storage options.’) under the ‘Data Storage’ section.

2. There are external links embedded in the content. Do we want to include them under the ‘References’ section as well? Will this be seen as repeated links? Same comment for ‘Enhanced Workflow’.

3. Would including links (under the ‘References’ section) to REST and gRPC be useful to the readers?

Enhanced Workflow

4. It seems like we did not cover ‘Experiment & Pipeline Tracking’. Do we want to?

**Comment by Syakyr** [comment]
> 1. A hyperlink seems to be missing in this sentence (‘You can check this section for more information on data storage options.’) under the ‘Data Storage’ section.

Fixed. It's for linking to #30, but I forgot to update that. 

> 2. There are external links embedded in the content. Do we want to include them under the ‘References’ section as well? Will this be seen as repeated links? Same comment for ‘Enhanced Workflow’.

I've repeated links with #30, so I'll do that here as well. Markdown supports reference linking, so it's not much of a hassle to change, or rather, it looks cleaner in code.

> 3. Would including links (under the ‘References’ section) to REST and gRPC be useful to the readers?

Because the handbook is meant to be agnostic in its solutions, I've only put REST and gRPC protocols as examples, of which I would assume that the reader would go read further on what those are, and discover what wrappers are available to set up those protocols with their programming language of choice. We could improve on this, but I would prefer to get some feedback that they would need more guidance with regards to how those protocols work, with Python as an example, for instance, or how those protocols interact with other programs or the user I suppose.

> 4. It seems like we did not cover ‘Experiment & Pipeline Tracking’. Do we want to?

Seems that I've missed that section while splitting the articles into two. I've added that together with the Model Monitoring section since it's quite similar while targeting different processes.

File: book/5-data-mgmt-exp-proc/e2e-workflow.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+Most, if not all, codebases benefit from version control to reduce 
+errors by rolling back changes to a known working source. You may 
+already have been exposed to it through GitHub or GitLab, which are the
+leading platforms for source code management especially for open source
+projects. Understanding how to use version control effectively would 
+make the end-to-end workflow easier.
+
+For AI practitioners such as ourselves, there is also a subset of 
+version control that targets toward datasets, which you may read 
+[this article][data-versioning] for more information.
```
Comment (kevinosl) [inline book/5-data-mgmt-exp-proc/e2e-workflow.md:116]:
```which you may read 
[this article][data-versioning] for more information.```

Comment (kevinosl) [reply book/5-data-mgmt-exp-proc/e2e-workflow.md:116]: I think the link to the article still appears as a placeholder. May wish to check.

File: book/5-data-mgmt-exp-proc/e2e-workflow.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
+# What are the Processes Involved to Build a Basic End-to-End Workflow?
```
Comment (kevinosl) [inline book/5-data-mgmt-exp-proc/e2e-workflow.md]: Would this sound better?
# What are the Processes Involved in Building a Basic End-to-End Workflow?

**Comment by kevinosl** [comment]
General comment: were the diagram(s) created ourselves, or taken from a framework? If its the latter, it would be good to have a caption to reference the framework and also place the link in the References.

If its the former, it might also still be good to add some references to industry-accepted workflow frameworks.
(PS: by workflow frameworks I am referring not to TDSP or CRISP-DM, which personally I consider to be more of high-level lifecycle frameworks rather than workflow frameworks. Those may sit well in the PM chapter instead. Not sure if you guys have a different view.)

Lastly, it might be good to color code the basic and enhanced nodes in the combined diagram, so readers can orientate themselves more easily at one glance. This may be especially helpful in the Enhanced article, which only discusses the nodes that are not captured in the Basic article.

File: book/5-data-mgmt-exp-proc/e2e-workflow.md (hunk 1 above)
Comment (Syakyr) [reply book/5-data-mgmt-exp-proc/e2e-workflow.md:116]: I've fixed this issue; it seems that there's a typo on line 155 that breaks the hyperlinks.

File: book/5-data-mgmt-exp-proc/e2e-workflow.md (hunk 2 above)
Comment (Syakyr) [reply book/5-data-mgmt-exp-proc/e2e-workflow.md]: Yup. I've changed the title to the recommended one 👍🏼

**Comment by Syakyr** [comment]
> General comment: were the diagram(s) created ourselves, or taken from a framework? If its the latter, it would be good to have a caption to reference the framework and also place the link in the References.
> 
> If its the former, it might also still be good to add some references to industry-accepted workflow frameworks. (PS: by workflow frameworks I am referring not to TDSP or CRISP-DM, which personally I consider to be more of high-level lifecycle frameworks rather than workflow frameworks. Those may sit well in the PM chapter instead. Not sure if you guys have a different view.)

The diagram was created from Ryzal's slides with regards to our MLOps framework, but with the product names removed to ensure generality within the handbook. I've linked to the ml-ops.org website should readers want to know more with regards to end-to-end workflows and MLOps in general, although it's not just pertaining to the end-to-end workflow. I can give a direct link to the section on the workflows, but I would assume that they've read both sections and want to know more, and they'll go to the website linked at the end of the advanced workflows section, so it might be redundant to add references on the 1st section.

> Lastly, it might be good to color code the basic and enhanced nodes in the combined diagram, so readers can orientate themselves more easily at one glance. This may be especially helpful in the Enhanced article, which only discusses the nodes that are not captured in the Basic article.

I've used a different color shade to denote the advanced nodes in both sections to differentiate between the basic and enhanced nodes 👍🏼

**Comment by ryzalk** [comment]
Closing this pull request in view of the new merging workflow i.e. from contributing branches to `develop`.
CC @chuawjk

**Pull request closed without merging**